### PR TITLE
Feature: add reinforcement learning as an advanced option for training

### PR DESCRIPTION
**Pull request by hkunzhe**
### UI
image

### Results
0.56 -> 0.65
image

### CODE REVIEW

File: scripts/train_kohya/ddpo_pytorch/rewards.py
[the remark this answers is not in the document]
Comment (wuziheng) [reply scripts/train_kohya/ddpo_pytorch/rewards.py:1]: 以及函数comments 后续其实我们有出lib和自动化doc的计划，相对原创的code 请遵守好的coding规范，可以使用chatgpt生成可以自动导出sphnix 文档的函数注释

File: scripts/train_kohya/ddpo_pytorch/rewards.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+    # Redundant parameters are for backward compatibility.
+    def __call__(src_images: Union[List[Image.Image], np.array, torch.Tensor], prompts: List[str]) -> Tuple[np.array, dict]:
+        src_images = _convert_images(src_images)
+        src_embs = []  # (N, 512)
+        for s in src_images:
+            try:
+                emb = face_recognition(dict(user=s))[OutputKeys.IMG_EMBEDDING][0]
+            except:  # cv2.error; TypeError.
+                emb = np.array([0] * 512)
```
Comment (wuziheng) [inline scripts/train_kohya/ddpo_pytorch/rewards.py:67]: 吞掉了 原有的错误，其实可能还有别的错误，所以这里推荐的做法把error catch 成str 打出来

File: scripts/train_kohya/ddpo_pytorch/stat_tracking.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
+import numpy as np
+from collections import deque
+
+
+class PerPromptStatTracker:
+    def __init__(self, buffer_size, min_count):
```
Comment (wuziheng) [inline scripts/train_kohya/ddpo_pytorch/stat_tracking.py:6]: 请简单描述这个函数的功能和在算法中的作用 方便其余开发者结合wiki follow

**Review by wuziheng — COMMENTED** [review]
first view

File: scripts/train_kohya/ddpo_pytorch/rewards.py (hunk 1 above)
Comment (wuziheng) [reply scripts/train_kohya/ddpo_pytorch/rewards.py:67]: 举例子 可能会oom maybe？

File: scripts/train_kohya/ddpo_pytorch/rewards.py
[the remark this answers is not in the document]
Comment (hkunzhe) [reply scripts/train_kohya/ddpo_pytorch/rewards.py:1]: > facied_v0和v2过于潦草了，还是设计一下吧

这个你指的是函数名吗？

File: scripts/train_kohya/ddpo_pytorch/rewards.py (hunk 1 above)
Comment (hkunzhe) [reply scripts/train_kohya/ddpo_pytorch/rewards.py:67]: > 吞掉了 原有的错误，其实可能还有别的错误，所以这里推荐的做法把error catch 成str 打出来

这里我可以抛出，但是建议上面 modelscope 的 pipeline 抛出一下。因为换了 FaceID 的 backbone，导致一些（文生图生成的）侧脸照没法检测出人脸。所以这一块我都 set reward to 0 by default.

File: scripts/train_kohya/ddpo_pytorch/rewards.py
[the remark this answers is not in the document]
Comment (hkunzhe) [reply scripts/train_kohya/ddpo_pytorch/rewards.py:1]: > 以及函数comments 后续其实我们有出lib和自动化doc的计划，相对原创的code 请遵守好的coding规范，可以使用chatgpt生成可以自动导出sphnix 文档的函数注释

这个可以确认一个 python code style 吗，然后可以设置 pyproject.toml 来用 black 和 isort 自动化规范。

**Comment by hkunzhe** [comment]
https://github.com/aigc-apps/sd-webui-EasyPhoto/pull/109/commits/fb49bae1f16b0287bebab381201af3cda21db9fb does the following optimizations:
1. Update default hyper-parameters (e.g., learning rate and training batch size) for smaller computing resources and improve the training speed.
2. Fix the problem of not being able to stop training when the time runs out. @bubbliiiing help me test in Windows.
3. Code and docstring standards @wuziheng

**Comment by hkunzhe** [comment]
https://github.com/aigc-apps/sd-webui-EasyPhoto/pull/109/commits/dabb5c128d1e2648217cc90553b251ffb4b10dfb:
1. Use `thread.Timer` instead of `signal` to limit the execution time of a function call. It should work in Windows @bubbliiiing.
2. Save the best checkpoint (be copied to webui) and maintain a heap (except for epoch 0) with length `num_checkpoint_limit - 1` by `reward_mean`.

File: scripts/train_kohya/ddpo_pytorch/rewards.py
[the remark this answers is not in the document]
Comment (wuziheng) [reply scripts/train_kohya/ddpo_pytorch/rewards.py:1]: 可以的，我之前都是用pre-commit 约束的。但自动化doc的话我不确定你提供的方案能够约束，我都是casebycase 手动测试的。

File: scripts/train_kohya/ddpo_pytorch/rewards.py (hunk 1 above)
Comment (wuziheng) [reply scripts/train_kohya/ddpo_pytorch/rewards.py:85]: OK

**Comment by wuziheng** [comment]
已经完成测试，faceid 分数来到0.7+
zzmx

**Review by wuziheng — APPROVED** [review]
LGTM， This is a big feature, provide a method to improve Lora behavior according to some preset metric, should be notice and read by anyone who interested of further development of EP.

**Comment by wuziheng** [comment]
@hkunzhe  @bubbliiiing  Should Notice that 
During my testing, I found that the optimized model indeed showed a significant improvement in the similarity scores of the generated characters in end-to-end rendering, for example, increasing from an average of 0.55 to close to 0.7+. However, the visual similarity did not seem to improve as much, and there appeared to be some inconsistency in the identity across multiple renderings, giving the impression of a certain degree of overfitting. All of these aspects would require longer-term and larger-scale testing within the community to validate.

**Comment by hkunzhe** [comment]
From my experience, the gap between faceID scores and human perception may have the following reasons:
1. Insufficient network capabilities for extracting facial features: my initial test was based on [faceid_SCRFD](https://github.com/aigc-apps/sd-webui-EasyPhoto/blob/main/scripts/train_kohya/ddpo_pytorch/rewards.py#L27), and the scores would be more aligned with human perception. And during the test with [faceid_retina](https://github.com/aigc-apps/sd-webui-EasyPhoto/blob/main/scripts/train_kohya/ddpo_pytorch/rewards.py#L60), it was found that many face images directly generated by LoRA could not be recognized by the model, which rarely happened in [faceid_SCRFD](https://github.com/aigc-apps/sd-webui-EasyPhoto/blob/main/scripts/train_kohya/ddpo_pytorch/rewards.py#L27).
2. The training objectives of reinforcement learning and the actual inference pipeline are not completely aligned.
